ocean_lib/ocean/v4/ocean_assets.py

```diff
@@ -0,0 +1,408 @@
+
+        # copy metadata so we don't change the original
+        metadata_copy = copy.deepcopy(metadata)
+
+        asset_type = metadata_copy["type"]
+        assert asset_type in (
+            "dataset",
+            "algorithm",
+        ), f"Invalid/unsupported asset type {asset_type}"
+
+        if not provider_uri:
+            provider_uri = DataServiceProvider.get_url(self._config)
+
+        address = get_address_of_type(self._config, ERC721FactoryContract.CONTRACT_NAME)
+        erc721_factory = ERC721FactoryContract(self._web3, address)
+
+        if not erc721_address:
+            name = erc721_name or metadata["name"]
+            symbol = erc721_symbol or name
+            additional_erc20_deployer = erc721_additional_erc_deployer or ZERO_ADDRESS
+            token_uri = erc721_uri or "https://oceanprotocol.com/nft/"
+            # register on-chain
+            tx_id = erc721_factory.deploy_erc721_contract(
+                name,
+                symbol,
+                template_index=template_index,
+                additional_erc20_deployer=additional_erc20_deployer,
+                token_uri=token_uri,
+                from_wallet=publisher_wallet,
+            )
+            created = get_timestamp()
+            _ = self._web3.eth.wait_for_transaction_receipt(tx_id)
+            erc721_token = ERC721Token(
+                self._web3, erc721_factory.get_token_address(tx_id)
+            )
+            if not erc721_token:
+                logger.warning("Creating new data token failed.")
+                return None
+            logger.info(
+                f"Successfully created data token with address "
+                f"{erc721_token.address} for new dataset asset."
+            )
+            erc721_address = erc721_token.address
+        else:
+            # verify nft address
+            if not erc721_factory.verify_nft(erc721_address):
+                raise ContractNotFound(
+                    f"NFT address {erc721_address} is not found in the ERC721Factory events."
+                )
+
+        assert erc721_address, "nft_address is required for publishing a dataset asset."
+        erc721_token = ERC721Token(self._web3, erc721_address)
+
+        # Create a DDO object
+        asset = V4Asset()
+
+        # Generating the did and adding to the ddo.
+        did = f"did:op:{create_checksum(erc721_token.address + str(self._web3.eth.chain_id))}"
+        asset.did = did
+        # Check if it's already registered first!
+        if self._get_aquarius().ddo_exists(did):
+            raise AquariusError(
+                f"Asset id {did} is already registered to another asset."
+            )
+        asset.chain_id = self._web3.eth.chain_id
+        asset.metadata = metadata
+
+        asset.credentials = credentials if credentials else build_credentials_dict()
+
+        erc20_addresses = []
+        services = services or []
+        deployed_erc20_tokens = deployed_erc20_tokens or []
```

Comment:
I'm sorry but this is very confusing and I'm not sure what is happening here. I do not understand the connection between the tokens and the services. Why are we "guessing" correspondences? Why do we have a list of data_tokens which is different from the deployed list? I know we want to allow for multiple options but I'm not sure this is it. And nobody seems to know what we are trying to build.

The basic flow with one datatoken, one access service? Ok, we are customising this a lot, but we are not organised in any way. @AnaLoznianu who should I ask for help here? This is beyond my knowledge of what ocean.py should do. I need help ASAP.